ruff_lsp/server.py

```diff
@@ -397,6 +408,35 @@ async def hover(params: HoverParams) -> Hover | None:
     return None
 
 
+@LSP_SERVER.feature(TEXT_DOCUMENT_DOCUMENT_LINK)
+async def document_link(params: DocumentLinkParams) -> list[DocumentLink] | None:
+    """LSP handler for textDocument/documentLink request."""
+    document = LSP_SERVER.workspace.get_document(params.text_document.uri)
+
+    links: list[DocumentLink] = []
+    for index, line in enumerate(document.lines):
```

Comment:
Oh, this request applies for an _entire_ document and not for a specific range. It makes me wonder how often this request is being sent by the client. If it's being sent multiple times then we should think of ways to optimize this as otherwise this might take some time for a large file irrespective of whether it contains the codes or not.

A few options might be:
* Applying the regex on the entire document using [`re.finditer`](https://docs.python.org/3/library/re.html?highlight=re%20finditer#re.finditer) instead of applying the regex on each line
* Using [`documentLink/resolve`](https://docs.python.org/3/library/re.html?highlight=re%20finditer#re.finditer) to lazily compute the rule name. So, for `textDocument/documentLink` request, we would only compute the range and when the client explicitly requests for the link, we would run the command and update the `documentLink` object.

Reply:
@dhruvmanila Thanks for the comment!

> It makes me wonder how often this request is being sent by the client.

- When a file is opened.
- When a file is saved.

Reply:
`textDocument/documentLink` is sent on type.

Reply:
I'll close this :) Iterating over the entire document on type is too expensive.